File: libs/semester.ts
export enum SemesterCode {
  jTerm = "ja",
  spring = "sp",
  summer = "su",
  fall = "fa",
}

export type SemesterInfo = { semesterCode: SemesterCode; year: number };

export default class Semester {
  private static readonly semesterCodes = [
    SemesterCode.jTerm,
    SemesterCode.spring,
    SemesterCode.summer,
    SemesterCode.fall,
  ];

  private static readonly numOfSemesters = Semester.semesterCodes.length;

  readonly semesterCode: SemesterCode;
  readonly year: number;

  constructor(
    {
      semesterCode,
      year,
    }: SemesterInfo = Semester.predictCurrentSemester().toJSON()
  ) {
    this.semesterCode = semesterCode;
    this.year = Math.floor(year);
  }

  getSemesterName() {
    return {
      [SemesterCode.jTerm]: "J-Term",
      [SemesterCode.spring]: "Spring",
      [SemesterCode.summer]: "Summer",
      [SemesterCode.fall]: "Fall",
    }[this.semesterCode];
  }

  toCode() {
    return `${this.semesterCode}${this.year}`;
  }

  toString() {
    return `${this.getSemesterName()} ${this.year}`;
  }

  toJSON() {
    return { semesterCode: this.semesterCode, year: this.year };
  }

  validate() {
    return Semester.getSemesterOptions().some((semester) =>
      Semester.equals(semester, this)
    );
  }

  static fromCode(code?: string) {
    const semesterCode = code?.slice(0, 2)?.toLowerCase() as SemesterCode;
    const year = parseInt(code?.slice(2) || "");

    if (year && this.semesterCodes.includes(semesterCode)) {
      const semester = new Semester({ semesterCode, year });
      if (semester.validate()) {
        return semester;
      }
    }

    return undefined;
  }

  static predictCurrentSemester() {
    const today = new Date();
    const month = today.getMonth() + 1;

    let semesterCode;

    if (month <= 1) {
      semesterCode = SemesterCode.jTerm;
    } else if (month <= 5) {
      semesterCode = SemesterCode.spring;
    } else if (month <= 8) {
      semesterCode = SemesterCode.summer;
    } else {
      semesterCode = SemesterCode.fall;
    }

    return new Semester({ semesterCode, year: today.getFullYear() });
  }

  static predictFurthestSemester() {
    const today = new Date();
    const month = today.getMonth() + 1;
    let year = today.getFullYear();

    let semesterCode;

    if (month < 3) {
      semesterCode = SemesterCode.summer;
    } else if (month < 9) {
      semesterCode = SemesterCode.fall;
    } else if (month < 10) {
      semesterCode = SemesterCode.jTerm;
      year += 1;
    } else {
      semesterCode = SemesterCode.spring;
      year += 1;
    }

    return new Semester({ semesterCode, year });
  }

  static getSemesterOptions(showNext: boolean = true, prevCount: number = 4) {
    const semesterOptions = [];
    const current = Semester.predictCurrentSemester();
    let start = current.prev(Math.max(0, prevCount));

    const end = (
      showNext ? Semester.predictFurthestSemester() : current
    ).next();

    while (!Semester.equals(start, end)) {
      semesterOptions.push(start);
      start = start.next();
    }

    return semesterOptions;
  }

  static equals(lhs: SemesterInfo, rhs: SemesterInfo) {
    return lhs.semesterCode === rhs.semesterCode && lhs.year === rhs.year;
  }

  static between(lhs: SemesterInfo, rhs: SemesterInfo) {
    return (
      (lhs.year - rhs.year) * Semester.numOfSemesters +
      Semester.semesterCodes.indexOf(lhs.semesterCode) -
      Semester.semesterCodes.indexOf(rhs.semesterCode)
    );
  }

  prev(n: number = 1) {
    return this.next(-n);
  }

  next(n: number = 1) {
    let index =
      Semester.semesterCodes.indexOf(this.semesterCode) + Math.floor(n);
    const year = this.year + Math.floor(index / Semester.numOfSemesters);
    index =
      ((index % Semester.numOfSemesters) + Semester.numOfSemesters) %
      Semester.numOfSemesters;

    return new Semester({ semesterCode: Semester.semesterCodes[index], year });
  }
}

File: ios/RateMyClasses Watch App/Semester.swift
//
//  Semester.swift
//  RateMyClasses Watch App
//
//  Created by McCoy Zhu on 11/30/22.
//

import Foundation

struct Semester: Codable, Equatable, Comparable {
  enum SemesterCode: String, Codable, CaseIterable {
    case ja, sp, su, fa
  }
  
  let semesterCode: SemesterCode
  let year: Int
  
  var code: String {
    return "\(semesterCode.rawValue)\(year)"
  }
  
  var name: String {
    let semesterName: String
    switch semesterCode {
      case .ja:
        semesterName = "J-Term"
      case .sp:
        semesterName = "Spring"
      case .su:
        semesterName = "Summer"
      case .fa:
        semesterName = "Fall"
    }
    return "\(semesterName) \(year)"
  }
  
  static func predictCurrentSemester() -> Semester {
    let semesterCode: SemesterCode
    let month = Calendar(identifier: .gregorian).component(.month, from: Date())
    let year = Calendar(identifier: .gregorian).component(.year, from: Date())
    
    if (month <= 1) {
      semesterCode = .ja
    } else if (month <= 5) {
      semesterCode = .sp
    } else if (month <= 8) {
      semesterCode = .su
    } else {
      semesterCode = .fa
    }
    
    return Semester(semesterCode: semesterCode, year: year)
  }
  
  static func -(lhs: Semester, rhs: Semester) -> Int {
    return (lhs.year - rhs.year) * 4 + (SemesterCode.allCases.firstIndex(of: lhs.semesterCode) ?? 0) - (SemesterCode.allCases.firstIndex(of: rhs.semesterCode) ?? 0)
  }
  
  static func <(lhs: Semester, rhs: Semester) -> Bool {
    return lhs - rhs < 0
  }
  
  static func >(lhs: Semester, rhs: Semester) -> Bool {
    return lhs - rhs > 0
  }
}
